### PR TITLE
Add an Option for Automatically Enabling NF Mod upon Failure

### PR DESCRIPTION
**Pull request by QiuFeng54321**
If one just wishes to enjoy the music and isn't sure if they would pass the map, but would want to:
+ Still be able to submit their scores if they don't die
+ Still be able to play even if they die without `NF`

Then this could be very helpful.

+ [x] Add relevant config and option item
+ [x] Add handling code for enabling NF
+ [x] Disable NF upon leaving when it had been enabled during gameplay
+ + [x] Works when `Skip Results Screen` is enabled
+ + [x] Works when retrying
+ + [x] Does not affect Calibration view
+ + [x] Does not clear the NF mod if it is intentionally activated by the player
+ [x] NoMiss mod would take priority, such that it would restart the game upon missing when AutoNoFail is set
+ [x] Would not affect multiplayer (Well I couldn't check this but there's logic blocking it)
+ [x] Find a suitable place for the option item (please help me devs, I can't find an appropriate category so I just put it below 'Tap to Pause' for now)

### CODE REVIEW

**Comment by QiuFeng54321** [comment]
Starting in [0a7b786](https://github.com/Quaver/Quaver/pull/3968/commits/0a7b7868edffb1d0bbde19366d2590bf8b2eaa08), a different logic is used, as described in the commit message. The reason for this change is that `Failed` is being used 20 times, and it being `true` for even a millisecond could potentially cause other code using this as a condition to trigger. Now, `Failed` is `false` when `AutoNoFail` is set and the player fails. A separate logic in `HandleFailure()` is used for the enabling of NF instead. This logic is the original logic of `Failed` but with some modifications. I need advises in potentially merging the two logic in some way of extracting some common part out of it, so that further modifications related to the failing condition would not need to take care of both logic at the same time.

File: Quaver.Shared/Screens/Gameplay/GameplayScreen.cs [hunk 1]
```diff
@@ -527,6 +536,8 @@ public override void Update(GameTime gameTime)
             QuaverScreenChangeType type = QuaverScreenChangeType.CompleteChange)
         {
             Utils.NativeUtils.EnableWindowsKey();
+            if (IsNoFailAddedInGameplay)
+                ModManager.RemoveMod(ModIdentifier.NoFail);
```
Comment (Warp9000) [inline Quaver.Shared/Screens/Gameplay/GameplayScreen.cs:540]: Removing NoFail here will make result screen try to submit the score online, you should remove it after it has checked for unranked mods.

File: Quaver.Shared/QuaverGame.cs [hunk 2]
```diff
@@ -947,7 +947,6 @@ public static void SetAudioDevice(bool reloadResources = false)
             {
                 if (ConfigManager.AudioOutputDevice.Value != Bass.GetDeviceInfo(i).Name)
                     continue;
-
```
Comment (Warp9000) [inline Quaver.Shared/QuaverGame.cs:950]: Whats with the random whitespace removal?

**Review by Warp9000 — CHANGES_REQUESTED** [review]

File: Quaver.Shared/QuaverGame.cs (hunk 2 above)
Comment (QiuFeng54321) [reply Quaver.Shared/QuaverGame.cs]: Fixed

File: Quaver.Shared/Screens/Gameplay/GameplayScreen.cs (hunk 1 above)
Comment (QiuFeng54321) [reply Quaver.Shared/Screens/Gameplay/GameplayScreen.cs]: Fixed. Now removes NF only when retrying, quitting from pause menu and after calling SubmitOnlineScore()

**Comment by AiAe** [comment]
I was discussing with @Warp9000 wont it be better if we add new option for this, instead applying the mod mid game?
As in, players can toggle if after they fail they want to continue playing, and then we can just apply the NF mod at the end of the play? 
Wont this solution be better and safer than what your doing?

**Comment by QiuFeng54321** [comment]
> I was discussing with @Warp9000 wont it be better if we add new option for this, instead applying the mod mid game? As in, players can toggle if after they fail they want to continue playing, and then we can just apply the NF mod at the end of the play? Wont this solution be better and safer than what your doing?

I changed it so that it applies the NF mod when the Results Screen loads and before the score submits. 

Also I made an option subcategory `Gameplay->Others` and put the toggle there. I don't really think it could be fit in anywhere.

**Comment by QiuFeng54321** [comment]
Closing due to too much commits containing abandoned features. Reopening a new one.

**Pull request closed without merging**